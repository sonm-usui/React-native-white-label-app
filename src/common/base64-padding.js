// @flow

export function addBase64Padding(data: string): string {
  const trail = data.length % 4
  if (trail === 0) {
    return data
  }

  return `${data}${'='.repeat(4 - trail)}`
}

export function removeBase64Padding(data: string): string {
  return data.split('=')[0]
}
