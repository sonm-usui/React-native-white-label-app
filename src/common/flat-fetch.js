// @flow

// If you look into the return type of this function
// you should notice that we are returning an array from  this function
// this function is just a wrapper on top of fetch.
// However, this changes a lot of ways we write code and handle errors
// By returning an array in which first item is an error
// and second item is the response
// We get two important things
// 1. We don't have to write try...catch statements everywhere
// 2. We are forced to handle errors. No more forgetting to add try...catch
// TODO:KS Write more details about both 1 & 2
// Problem #1 leads to try...catch nesting hell if there are multiple async
// calls
// Problem #2 leads to unexpected app crashes at runtime

export const acceptHeaders = { Accept: 'application/json;flavor=didcomm-msg' }

export async function flatFetch(
  url: string,
  body?: string,
  headers?: { [string]: any }
): Promise<[null | typeof Error, null | string, null | Response]> {
  const options = {
    mode: 'cors',
    ...(body ? { method: 'POST', body } : {}),
    ...(headers ? headers : {}),
  }
  try {
    const response = await fetch(url, options)
    const responseText = await response.text()
    // request redirected
    if (response && response.status === 302) {
      return [null, null, response]
    }
    if (!response.ok) {
      throw new Error(responseText)
    }
    return [null, responseText, response]
  } catch (e) {
    return [e, null, null]
  }
}
