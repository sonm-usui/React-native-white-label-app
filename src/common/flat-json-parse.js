// @flow

export function flatJsonParse(
  text: string
): [null | typeof Error, null | Object] {
  try {
    return [null, JSON.parse(text)]
  } catch (e) {
    return [e, null]
  }
}
