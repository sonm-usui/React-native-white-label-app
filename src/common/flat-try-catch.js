// @flow

export const flatTryCatch = (fn: (...variableArgs: any[]) => any) => (
  ...args: any[]
): [null | typeof Error, null | any] => {
  try {
    return [null, fn(...args)]
  } catch (e) {
    return [e, null]
  }
}
