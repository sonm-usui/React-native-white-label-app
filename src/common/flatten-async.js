// @flow

// If you look into the return type of this function
// you should notice that we are returning an array from  this function
// this is just a helper function to wrap on top of other async functions
// However, this changes a lot of ways we write code and handle errors
// By returning an array in which first item is an error
// and second item is the response
// We get two important things
// 1. We don't have to write try...catch statements everywhere
// 2. We are forced to handle errors. No more forgetting to add try...catch
// TODO:KS Write more details about both 1 & 2
// Here are the problems that this function aims to solve
// Problem #1 leads to try...catch nesting hell if there are multiple async
// calls
// Problem #2 leads to unexpected app crashes at runtime

export const flattenAsync = (
  fn: (...variableArgs: any[]) => Promise<any>
) => async (
  ...args: Array<any>
): Promise<[null | typeof Error, null | any]> => {
  try {
    const response = await fn(...args)
    return [null, response]
  } catch (e) {
    return [e, null]
  }
}
