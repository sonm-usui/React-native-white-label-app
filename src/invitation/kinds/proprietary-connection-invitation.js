import {
  QR_CODE_DELEGATION_DID,
  QR_CODE_DELEGATION_KEY,
  QR_CODE_DELEGATION_SIGNATURE,
  QR_CODE_LOGO_URL,
  QR_CODE_REQUEST_ID,
  QR_CODE_SENDER_AGENCY,
  QR_CODE_SENDER_AGENCY_DID,
  QR_CODE_SENDER_AGENCY_ENDPOINT,
  QR_CODE_SENDER_AGENCY_KEY,
  QR_CODE_SENDER_DETAIL,
  QR_CODE_SENDER_DID,
  QR_CODE_SENDER_KEY_DELEGATION,
  QR_CODE_SENDER_NAME,
  QR_CODE_SENDER_PUBLIC_DID,
  QR_CODE_SENDER_VERIFICATION_KEY,
  QR_CODE_TARGET_NAME,
  QR_CODE_VERSION
} from "../../api/api-constants";
import {schemaValidator} from "../../services/schema-validator";
import type {InvitationPayload, ShortProprietaryConnectionInvitation, ProprietaryConnectionInvitation} from "../type-invitation";

export function isShortProprietaryInvitation(
  data: Object
): ShortProprietaryConnectionInvitation | false {
  // check if qr code schema is valid
  if (!schemaValidator.validate(proprietaryShortInviteSchema, data)) {
    return false
  }

  // since all checks are passed
  // qr code data is valid, send the parsed data back
  return data
}

export function isProprietaryInvitation(
  data: Object
): ProprietaryConnectionInvitation | false {
  // check if qr code schema is valid
  if (!schemaValidator.validate(proprietaryInvitationSchema, data)) {
    return false
  }

  // since all checks are passed
  // qr code data is valid, send the parsed data back
  return data
}

export function convertShortProprietaryInvitationToAppInvitation(qrCode: ShortProprietaryConnectionInvitation) {
  const qrSenderDetail = qrCode[QR_CODE_SENDER_DETAIL]
  const qrSenderAgency = qrCode[QR_CODE_SENDER_AGENCY]
  const senderDetail = {
    name: qrSenderDetail[QR_CODE_SENDER_NAME],
    agentKeyDlgProof: {
      agentDID:
        qrSenderDetail[QR_CODE_SENDER_KEY_DELEGATION][QR_CODE_DELEGATION_DID],
      agentDelegatedKey:
        qrSenderDetail[QR_CODE_SENDER_KEY_DELEGATION][QR_CODE_DELEGATION_KEY],
      signature:
        qrSenderDetail[QR_CODE_SENDER_KEY_DELEGATION][
          QR_CODE_DELEGATION_SIGNATURE
          ],
    },
    DID: qrSenderDetail[QR_CODE_SENDER_DID],
    logoUrl: qrSenderDetail[QR_CODE_LOGO_URL],
    verKey: qrSenderDetail[QR_CODE_SENDER_VERIFICATION_KEY],
    publicDID: qrSenderDetail[QR_CODE_SENDER_PUBLIC_DID],
  }

  const senderAgencyDetail = {
    DID: qrSenderAgency[QR_CODE_SENDER_AGENCY_DID],
    verKey: qrSenderAgency[QR_CODE_SENDER_AGENCY_KEY],
    endpoint: qrSenderAgency[QR_CODE_SENDER_AGENCY_ENDPOINT],
  }

  return {
    senderEndpoint: senderAgencyDetail.endpoint,
    requestId: qrCode[QR_CODE_REQUEST_ID],
    senderAgentKeyDelegationProof: senderDetail.agentKeyDlgProof,
    senderName: senderDetail.name,
    senderDID: senderDetail.DID,
    senderLogoUrl: senderDetail.logoUrl,
    senderVerificationKey: senderDetail.verKey,
    targetName: qrCode[QR_CODE_TARGET_NAME],
    senderDetail,
    senderAgencyDetail,
    version: qrCode[QR_CODE_VERSION],
  }
}

export const convertProprietaryInvitationToAppInvitation = (
  pendingInvitation: ProprietaryConnectionInvitation
): InvitationPayload => ({
  senderEndpoint: pendingInvitation.senderAgencyDetail.endpoint,
  requestId: pendingInvitation.connReqId,
  senderAgentKeyDelegationProof:
  pendingInvitation.senderDetail.agentKeyDlgProof,
  senderName: pendingInvitation.senderDetail.name,
  senderDID: pendingInvitation.senderDetail.DID,
  senderLogoUrl: pendingInvitation.senderDetail.logoUrl,
  senderVerificationKey: pendingInvitation.senderDetail.verKey,
  targetName: pendingInvitation.targetName,
  senderAgencyDetail: pendingInvitation.senderAgencyDetail,
  senderDetail: pendingInvitation.senderDetail,
  version: pendingInvitation.version,
})

export const proprietaryShortInviteSchema = {
  type: 'object',
  properties: {
    [QR_CODE_REQUEST_ID]: {type: 'string'},
    [QR_CODE_SENDER_DETAIL]: {
      type: 'object',
      properties: {
        [QR_CODE_SENDER_DID]: {type: 'string'},
        [QR_CODE_SENDER_NAME]: {type: 'string'},
        [QR_CODE_LOGO_URL]: {type: 'string'},
        [QR_CODE_SENDER_VERIFICATION_KEY]: {type: 'string'},
        [QR_CODE_SENDER_KEY_DELEGATION]: {
          type: 'object',
          properties: {
            [QR_CODE_DELEGATION_DID]: {type: 'string'},
            [QR_CODE_DELEGATION_KEY]: {type: 'string'},
            [QR_CODE_DELEGATION_SIGNATURE]: {type: 'string'},
          },
          required: [
            QR_CODE_DELEGATION_DID,
            QR_CODE_DELEGATION_KEY,
            QR_CODE_DELEGATION_SIGNATURE,
          ],
        },
      },
      required: [
        QR_CODE_SENDER_DID,
        QR_CODE_SENDER_NAME,
        QR_CODE_LOGO_URL,
        QR_CODE_SENDER_VERIFICATION_KEY,
        QR_CODE_SENDER_KEY_DELEGATION,
      ],
    },
    [QR_CODE_SENDER_AGENCY]: {
      type: 'object',
      properties: {
        [QR_CODE_SENDER_AGENCY_DID]: {type: 'string'},
        [QR_CODE_SENDER_AGENCY_KEY]: {type: 'string'},
        [QR_CODE_SENDER_AGENCY_ENDPOINT]: {type: 'string'},
      },
      required: [
        QR_CODE_SENDER_AGENCY_DID,
        QR_CODE_SENDER_AGENCY_KEY,
        QR_CODE_SENDER_AGENCY_ENDPOINT,
      ],
    },
    [QR_CODE_TARGET_NAME]: {type: 'string'},
  },
  required: [
    QR_CODE_REQUEST_ID,
    QR_CODE_SENDER_DETAIL,
    QR_CODE_SENDER_AGENCY,
    QR_CODE_TARGET_NAME,
  ],
}

export const proprietaryInvitationSchema = {
  type: 'object',
  properties: {
    connReqId: { type: 'string' },
    statusCode: { type: 'string' },
    senderDetail: {
      type: 'object',
      properties: {
        name: { type: 'string' },
        agentKeyDlgProof: {
          type: 'object',
          properties: {
            agentDID: { type: 'string' },
            agentDelegatedKey: { type: 'string' },
            signature: { type: 'string' },
          },
          required: ['agentDID', 'agentDelegatedKey', 'signature'],
        },
        DID: { type: 'string' },
        logoUrl: { type: 'string' },
        verKey: { type: 'string' },
        publicDID: { type: 'string' },
      },
      required: ['name', 'agentKeyDlgProof', 'DID', 'logoUrl', 'verKey'],
    },
    senderAgencyDetail: {
      type: 'object',
      properties: {
        DID: { type: 'string' },
        verKey: { type: 'string' },
        endpoint: { type: 'string' },
      },
      required: [],
    },
    targetName: { type: 'string' },
    version: { type: 'string' },
  },
  required: ['senderDetail', 'senderAgencyDetail', 'targetName'],
}
