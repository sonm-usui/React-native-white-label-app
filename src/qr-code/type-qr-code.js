// @flow
import type { ReactNavigation } from '../common/type-common'
import { proofRequestReceived } from '../proof-request/proof-request-store'
import type { InvitationPayload } from '../invitation/type-invitation'
import { handleInvitation } from '../invitation/invitation-store'
import { openIdConnectUpdateStatus } from '../open-id-connect/open-id-connect-actions'
import { changeEnvironmentUrl } from '../switch-environment/switсh-environment-store'
import { claimOfferReceived } from '../claim-offer/claim-offer-store'

export type QRCodeScannerScreenState = {
  isCameraEnabled: boolean,
  appState: ?string,
  permission: boolean,
}

export type QRCodeScannerScreenProps = {
  currentScreen: string,
  openIdConnectUpdateStatus: typeof openIdConnectUpdateStatus,
  changeEnvironmentUrl: typeof changeEnvironmentUrl,
  handleInvitation: typeof handleInvitation,
  proofRequestReceived: typeof proofRequestReceived,
  claimOfferReceived: typeof claimOfferReceived,
  scanQrClose: () => void
} & ReactNavigation

export type OutOfBandNavigation = {
  mainRoute: string,
  backRedirectRoute: string,
  uid: string,
  invitationPayload: InvitationPayload,
  senderName: string,
}

export const MESSAGE_NO_CAMERA_PERMISSION = 'No Camera permission'

export const MESSAGE_ALLOW_CAMERA_PERMISSION =
  'Please allow connect me to access camera from camera settings'
