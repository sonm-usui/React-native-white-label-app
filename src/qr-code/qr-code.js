// @flow
import React, { Component } from 'react'
import { Alert, AppState, Platform, PermissionsAndroid } from 'react-native'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import { Container, QRScanner } from '../components'
import { color } from '../common/styles/constant'
import { handleInvitation } from '../invitation/invitation-store'
import {
  claimOfferRoute,
  openIdConnectRoute,
  proofRequestRoute,
  pushNotificationPermissionRoute,
  qrCodeScannerTabRoute,
} from '../common/'

import type { OIDCAuthenticationRequest } from '../components/qr-scanner/type-qr-scanner'
import type { Store } from '../store/type-store'
import type {
  AriesConnectionInvite,
  AriesOutOfBandInvite,
  InvitationPayload,
  ShortProprietaryConnectionInvitation,
} from '../invitation/type-invitation'
import type {
  QRCodeScannerScreenProps,
  QRCodeScannerScreenState,
} from './type-qr-code'
import type { QrCodeEphemeralProofRequest } from '../proof-request/type-proof-request'
import { withStatusBar } from '../components/status-bar/status-bar'
import {
  OPEN_ID_CONNECT_STATE,
  openIdConnectUpdateStatus,
} from '../open-id-connect/open-id-connect-actions'
import { convertShortProprietaryInvitationToAppInvitation } from '../invitation/kinds/proprietary-connection-invitation'
import { convertAriesOutOfBandInvitationToAppInvitation } from '../invitation/kinds/aries-out-of-band-invitation'
import { convertAriesInvitationToAppInvitation } from '../invitation/kinds/aries-connection-invitation'
import { proofRequestReceived } from '../proof-request/proof-request-store'

import { SCAN_QR_CLOSE_X_BUTTON } from '../feedback/log-to-apptentive'
import type { QrCodeEphemeralCredentialOffer } from '../claim-offer/ephemeral-claim-offer'
import { claimOfferReceived } from '../claim-offer/claim-offer-store'
import { getPushNotificationAuthorizationStatus } from '../push-notification/components/push-notification-permission-screen'
import { usePushNotifications } from '../external-imports'

export class QRCodeScannerScreen extends Component<
  QRCodeScannerScreenProps,
  QRCodeScannerScreenState
> {
  state = {
    appState: AppState.currentState,
    isCameraEnabled: true,
    permission: false,
  }

  subscription = null;
  permissionCheckIntervalId: ?IntervalID = undefined

  onShortProprietaryInvitationRead = async (
    qrCode: ShortProprietaryConnectionInvitation
  ) => {
    if (this.props.currentScreen === qrCodeScannerTabRoute) {
      this.props.handleInvitation(
        convertShortProprietaryInvitationToAppInvitation(qrCode)
      )
    }
  }

  // Please do not remove below commented lines
  // We will want to have environment switcher from QR code
  // sometime in future

  // onAllowSwitchEnvironment = (url: EnvironmentSwitchUrlQrCode) => {
  //   this.props.changeEnvironmentUrl(url.url)
  //   this.props.navigation.goBack(null)
  // }

  // onEnvironmentSwitchUrl = (url: EnvironmentSwitchUrlQrCode) => {
  //   if (this.props.currentScreen === qrCodeScannerTabRoute) {
  //     Alert.alert(MESSAGE_RESET_CONNECT_ME, MESSAGE_RESET_DETAILS(url.name), [
  //       { text: 'Cancel' },
  //       {
  //         text: 'Switch',
  //         onPress: () => this.onAllowSwitchEnvironment(url),
  //       },
  //     ])
  //   }
  // }

  onProprietaryInvitationRead = async (payload: InvitationPayload) => {
    if (this.props.currentScreen === qrCodeScannerTabRoute) {
      this.props.handleInvitation(payload)
    }
  }

  onClose = () => {
    this.props.navigation.goBack(null)
    this.props.scanQrClose()
  }

  componentDidMount() {
    if (Platform.OS === 'android') {
      PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.CAMERA).then(
        (permission) => {
          if (permission === 'never_ask_again' || permission === 'denied') {
            this.props.navigation.goBack()
            Alert.alert(
              'Camera Permission Needed',
              'Please go into your device settings and enable camera permissions for Connect.Me to use the camera feature.',
              [
                {
                  text: 'OK',
                },
              ]
            )
            this.setState({ permission: false })
          }
          this.setState({ permission: true })
        }
      )
    }

    this.subscription = AppState.addEventListener('change', this._handleAppStateChange)
  }

  componentWillUnmount() {
    this.subscription.remove()
  }

  _handleAppStateChange = (nextAppState) => {
    if (
      this.state.appState &&
      this.state.appState.match(/inactive|background/) &&
      nextAppState === 'active'
    ) {
      this.setState({ isCameraEnabled: true })
    } else if (
      this.state.appState &&
      nextAppState.match(/inactive|background/) &&
      this.state.appState === 'active'
    ) {
      this.setState({ isCameraEnabled: false })
    }
    this.setState({ appState: nextAppState })
  }

  render() {
    return (
      <Container dark collapsable={true}>
        {this.state.isCameraEnabled &&
        this.props.navigation.isFocused() &&
        (this.state.permission || Platform.OS === 'ios') ? (
          <QRScanner
            onShortProprietaryInvitationRead={
              this.onShortProprietaryInvitationRead
            }
            onProprietaryInvitationRead={this.onProprietaryInvitationRead}
            onOIDCAuthenticationRequest={this.onOIDCAuthenticationRequest}
            onAriesConnectionInviteRead={this.onAriesConnectionInviteRead}
            onAriesOutOfBandInviteRead={this.onAriesOutOfBandInviteRead}
            onEphemeralProofRequest={this.onEphemeralProofRequest}
            onEphemeralCredentialOffer={this.onEphemeralCredentialOffer}
            onClose={this.onClose}
          />
        ) : null}
      </Container>
    )
  }

  onOIDCAuthenticationRequest = (
    oidcAuthenticationRequest: OIDCAuthenticationRequest
  ) => {
    this.props.openIdConnectUpdateStatus(
      oidcAuthenticationRequest,
      OPEN_ID_CONNECT_STATE.REQUEST_RECEIVED
    )
    this.props.navigation.navigate(openIdConnectRoute, {
      oidcAuthenticationRequest,
    })
  }

  onAriesConnectionInviteRead = async (
    ariesConnectionInvite: AriesConnectionInvite
  ) => {
    const invitation = convertAriesInvitationToAppInvitation(
      ariesConnectionInvite
    )
    this.props.handleInvitation(invitation)
  }

  onAriesOutOfBandInviteRead = async (invite: AriesOutOfBandInvite) => {
    const invitation = await convertAriesOutOfBandInvitationToAppInvitation(
      invite
    )
    this.props.handleInvitation(invitation)
  }

  onEphemeralProofRequest = (
    ephemeralProofRequest: QrCodeEphemeralProofRequest
  ) => {
    const uid = ephemeralProofRequest.ephemeralProofRequest['@id']
    this.props.proofRequestReceived(ephemeralProofRequest.proofRequestPayload, {
      uid,
      remotePairwiseDID:
        ephemeralProofRequest.ephemeralProofRequest['~service']
          .recipientKeys[0],
      senderLogoUrl: null,
      hidden: true,
    })
    this.props.navigation.navigate(proofRequestRoute, {
      uid,
      backRedirectRoute: this.props.route.params?.backRedirectRoute,
    })
  }

  handleiOSNavigation = async ({
                                       mainRoute,
                                       backRedirectRoute,
                                       uid,
                                       senderDID,
                                     }: any) => {
    const { navigation } = this.props
    const isAuthorized = await getPushNotificationAuthorizationStatus()
    const navigationFn = navigation.push || navigation.navigate
    const intendedPayload = {
      backRedirectRoute,
      uid,
      hidden: true,
    }

    if (Platform.OS === 'ios' && usePushNotifications && !isAuthorized) {
      navigationFn(pushNotificationPermissionRoute, {
        senderDID,
        navigatedFrom: qrCodeScannerTabRoute,
        intendedRoute: mainRoute,
        intendedPayload,
      })
    } else {
      navigationFn(mainRoute, intendedPayload)
    }
  }


  onEphemeralCredentialOffer = async (
    ephemeralCredentialOffer: QrCodeEphemeralCredentialOffer
  ) => {
    const uid = ephemeralCredentialOffer.ephemeralCredentialOffer['@id']
    const remotePairwiseDID = ephemeralCredentialOffer.ephemeralCredentialOffer['~service']
      .recipientKeys[0]

    this.props.claimOfferReceived(ephemeralCredentialOffer.claimOfferPayload, {
      uid,
      remotePairwiseDID,
      senderLogoUrl: null,
      hidden: true,
    })

    await this.handleiOSNavigation({
      mainRoute: claimOfferRoute,
      backRedirectRoute: this.props.route.params?.backRedirectRoute,
      uid,
      senderDID: remotePairwiseDID,
    })
  }
}

const mapStateToProps = (state: Store) => ({
  currentScreen: state.route.currentScreen,
})

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      openIdConnectUpdateStatus,
      handleInvitation,
      proofRequestReceived,
      claimOfferReceived,
      scanQrClose: () => SCAN_QR_CLOSE_X_BUTTON,
    },
    dispatch
  )

export const qrCodeScannerScreen = {
  routeName: qrCodeScannerTabRoute,
  screen: withStatusBar({ color: color.bg.sixth.color })(
    connect(mapStateToProps, mapDispatchToProps)(QRCodeScannerScreen)
  ),
}
