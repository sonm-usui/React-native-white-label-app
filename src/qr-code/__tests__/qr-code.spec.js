// @flow
import React from 'react'
import 'react-native'
import { Alert, Platform } from 'react-native'
import renderer from 'react-test-renderer'

import { qrCodeScannerTabRoute } from '../../common/'
import { QRCodeScannerScreen } from '../qr-code'
import {
  getNavigation,
  qrData,
  validQrCodeEnvironmentSwitchUrl,
} from '../../../__mocks__/static-data'
import { convertShortProprietaryInvitationToAppInvitation } from '../../invitation/kinds/proprietary-connection-invitation'

describe('<QRScannerScreen />', () => {
  function getProps() {
    return {
      navigation: getNavigation(),
      route: {},
      handleInvitation: jest.fn(),
      currentScreen: qrCodeScannerTabRoute,
      changeEnvironmentUrl: jest.fn(),
      enableCamera: jest.fn(),
      isCameraEnabled: true,
      openIdConnectUpdateStatus: jest.fn(),
      proofRequestReceived: jest.fn(),
      claimOfferReceived: jest.fn(),
      proofProposalReceived: jest.fn(),
      getPushNotificationAuthorizationStatus: jest.fn(),
      scanQrClose: jest.fn(),
    }
  }

  function setup() {
    const props = getProps()
    const component = renderer.create(<QRCodeScannerScreen {...props} />)
    const instance: QRCodeScannerScreen = component.getInstance()

    return { props, component, instance }
  }

  it('should match snapshot for ios platform', () => {
    const existingOS = Platform.OS
    Platform.OS = 'ios'

    const { instance, component } = setup()
    instance.setState({ isCameraEnabled: true })
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    Platform.OS = existingOS
  })

  it('should match snapshot for android platform', () => {
    const existingOS = Platform.OS
    Platform.OS = 'android'

    const { instance, component } = setup()
    instance.setState({ isCameraEnabled: true })
    instance.setState({ permission: true })
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    Platform.OS = existingOS
  })

  it('match snapshot when camera is not authorized for ios platform', () => {
    const existingOS = Platform.OS
    Platform.OS = 'ios'

    const { component } = setup()
    expect(component.toJSON()).toMatchSnapshot()

    Platform.OS = existingOS
  })

  it('match snapshot when camera is not authorized for android platform', () => {
    const existingOS = Platform.OS
    Platform.OS = 'android'

    const { component, instance } = setup()
    instance.setState({ permission: true })
    expect(component.toJSON()).toMatchSnapshot()

    Platform.OS = existingOS
  })

  it('should convert qr code to invitation', () => {
    expect(
      convertShortProprietaryInvitationToAppInvitation(qrData)
    ).toMatchSnapshot()
  })

  it('should redirect user to invitation screen on success read', () => {
    const {
      instance,
      props: { handleInvitation },
    } = setup()

    instance.onShortProprietaryInvitationRead(qrData)
    expect(handleInvitation).toHaveBeenCalledWith(
      convertShortProprietaryInvitationToAppInvitation(qrData)
    )
  })

  it('should navigate back to home if qr code scanner is closed', () => {
    const { instance, props } = setup()
    instance.onClose()
    expect(props.navigation.goBack).toHaveBeenCalledWith(null)
  })

  // skipping this test because as of now this code is commented and not used
  xit(`show alert if environment switch url is scanned,
      trigger action when Switch is clicked,
      and redirect to home tab`, () => {
    const {
      props: {
        changeEnvironmentUrl,
        navigation: { goBack },
      },
    } = setup()
    const alertSpy = jest.spyOn(Alert, 'alert')

    //instance.onEnvironmentSwitchUrl(environmentSwitchQrCodeData)
    expect(alertSpy).toHaveBeenCalled()

    const switchButton = alertSpy.mock.calls[0][2][1]
    // click switch button
    switchButton.onPress()

    expect(changeEnvironmentUrl).toHaveBeenCalledWith(
      validQrCodeEnvironmentSwitchUrl
    )

    expect(goBack).toHaveBeenCalledWith(null)

    alertSpy.mockReset()
    alertSpy.mockRestore()
  })
})
