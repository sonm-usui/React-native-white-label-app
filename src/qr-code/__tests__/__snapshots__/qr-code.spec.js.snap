// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QRScannerScreen /> match snapshot when camera is not authorized for android platform 1`] = `
<View
  collapsable={true}
  dark={true}
  style={
    Object {
      "backgroundColor": "#000000",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;

exports[`<QRScannerScreen /> match snapshot when camera is not authorized for ios platform 1`] = `
<View
  collapsable={true}
  dark={true}
  style={
    Object {
      "backgroundColor": "#000000",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;

exports[`<QRScannerScreen /> should convert qr code to invitation 1`] = `
Object {
  "requestId": "yta2odh",
  "senderAgencyDetail": Object {
    "DID": "BDSmVkzxRYGE4HKyMKxd1H",
    "endpoint": "52.38.32.107:80/agency/msg",
    "verKey": "6yUatReYWNSUfEtC2ABgRXmmLaxCyQqsjLwv2BomxsxD",
  },
  "senderAgentKeyDelegationProof": Object {
    "agentDID": "N2Uyi6SVsHZq1VWXuA3EMg",
    "agentDelegatedKey": "CTfF2sZ5q4oPcBvTP75pgx3WGzYiLSTwHGg9zUsJJegi",
    "signature": "/FxHMzX8JaH461k1SI5PfyxF5KwBAe6VlaYBNLI2aSZU3APsiWBfvSC+mxBYJ/zAhX9IUeTEX67fj+FCXZZ2Cg==",
  },
  "senderDID": "F2axeahCaZfbUYUcKefc3j",
  "senderDetail": Object {
    "DID": "F2axeahCaZfbUYUcKefc3j",
    "agentKeyDlgProof": Object {
      "agentDID": "N2Uyi6SVsHZq1VWXuA3EMg",
      "agentDelegatedKey": "CTfF2sZ5q4oPcBvTP75pgx3WGzYiLSTwHGg9zUsJJegi",
      "signature": "/FxHMzX8JaH461k1SI5PfyxF5KwBAe6VlaYBNLI2aSZU3APsiWBfvSC+mxBYJ/zAhX9IUeTEX67fj+FCXZZ2Cg==",
    },
    "logoUrl": "ent-logo-url",
    "name": "ent-name",
    "publicDID": undefined,
    "verKey": "74xeXSEac5QTWzQmh84JqzjuXc8yvXLzWKeiqyUnYokx",
  },
  "senderEndpoint": "52.38.32.107:80/agency/msg",
  "senderLogoUrl": "ent-logo-url",
  "senderName": "ent-name",
  "senderVerificationKey": "74xeXSEac5QTWzQmh84JqzjuXc8yvXLzWKeiqyUnYokx",
  "targetName": "there",
  "version": undefined,
}
`;

exports[`<QRScannerScreen /> should match snapshot for android platform 1`] = `
<View
  collapsable={true}
  dark={true}
  style={
    Object {
      "backgroundColor": "#000000",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;

exports[`<QRScannerScreen /> should match snapshot for ios platform 1`] = `
<View
  collapsable={true}
  dark={true}
  style={
    Object {
      "backgroundColor": "#000000",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;
